### PR TITLE
#1487 Fixed the ReqIF Exporter to not have a specobjectType for each requirement.

### LINKED ISSUE #1487: [Bug] When I export a reqIF it makes for every requirement a specObjectType instead of one.
When I export a reqIF it makes for every requirement a specObjectType instead of the logical amount of probably one or few.

### PR DESCRIPTION
**Pull request by rdevoogtatstariongroup**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-IME-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
#1487 Fixed the ReqIF Exporter to not have a specobjectType for each requirement. It makes a specObjectType per different combination of rules and simple parameter values.

### CODE REVIEW

File: Requirements.Tests/ReqIF/ParameterTypeMappingDialogTestFixture.cs [hunk 1]
```diff
@@ -180,6 +181,20 @@ public void VerifyThatMenuIsPopulated()
...
         }
 
+        [Test]
+        public void VerifyThatEnumValueRowsHaveNoCreateCommands()
+        {
+            var enumRow = this.dialog.MappingRows.First(x => x.Identifiable == this.enumDatadef);
+
+            // the enumeration datatype row itself can create an Enumeration Parameter Type
+            this.dialog.SelectedRow = enumRow;
+            Assert.AreEqual(1, this.dialog.CreateParameterTypeCommands.Count);
```
Comment (joao4all) [inline Requirements.Tests/ReqIF/ParameterTypeMappingDialogTestFixture.cs]: Suggestion: using Assert.That(list, Has.Count.EqualTo(...); is more idiomatic and follows the Constraint-Based Model for nunit tests.

This comment applies to every place where Assert.IsFalse/True/AreEqual/IsNotNull/etc... is used in this PR

File: Requirements/ReqIFDal/ReqIFBuilder.cs [hunk 2]
```diff
@@ -685,5 +723,45 @@ private void BuildToolExtension()
         {
             //TODO
         }
+
+        /// <summary>
+        /// Compares two sets of <see cref="ParameterizedCategoryRule"/>s for equality, ignoring order, so that
+        /// requirements can be grouped by their applied rule-set.
+        /// </summary>
+        private sealed class RuleSetEqualityComparer : IEqualityComparer<ParameterizedCategoryRule[]>
```
Comment (joao4all) [inline Requirements/ReqIFDal/ReqIFBuilder.cs]: Pls move this to its own file

File: Requirements/ReqIFDal/ThingFactory.cs [hunk 3]
```diff
@@ -680,5 +689,31 @@ private string GetAttributeValue(AttributeValue value)
...
         }
+
+        /// <summary>
+        /// Converts an XHTML (rich-text) attribute value to readable plain text. Line-breaking elements become new
+        /// lines, the remaining tags are removed and the XML/HTML entities are decoded. This keeps the textual content
+        /// of DOORS/Capella (and other) XHTML attributes instead of importing the raw markup.
+        /// </summary>
+        /// <param name="xhtml">The XHTML markup</param>
+        /// <returns>The plain-text representation</returns>
+        internal static string ConvertXhtmlToText(string xhtml)
```
Comment (joao4all) [inline Requirements/ReqIFDal/ThingFactory.cs]: 3 comments about this method
- Why specficially internal and not private?
- Regex is usually slow, and for this case, where I imagine this method will be used many times, maybe you should consider a different approach, like XmlReader (then HtmlDecode might be removed).
- If you decide to keep regex, please use it as a source generated fields. Also check for edge cases in this method tests, as we may encounter <style> or <script> tags (or more edgy things) in the xhtml, I suppose

File: Requirements/ViewModels/ReqIF/ReqIfExportDialogViewModel.cs [hunk 4]
```diff
@@ -149,11 +156,19 @@ public ReqIfExportDialogViewModel(IEnumerable<ISession> sessions, IEnumerable<It
                 this.Iterations.Add(new ReqIfExportIterationRowViewModel(iteration));
             }
 
+            this.WhenAnyValue(vm => vm.SelectedIteration).Subscribe(_ => this.PopulateRequirementsSpecifications());
+
+            this.WhenAnyValue(vm => vm.IncludeDeprecated).Subscribe(_ => this.PopulateRequirementsSpecifications());
+
```
Comment (joao4all) [inline Requirements/ViewModels/ReqIF/ReqIfExportDialogViewModel.cs]: The subscriptions in this class should be added to the disposables list

File: Requirements/ViewModels/ReqIF/ReqIfExportDialogViewModel.cs [hunk 5]
```diff
@@ -328,11 +382,108 @@ private async Task<ReqIF> BuildReqIf()
...
+            }
+
+            try
+            {
+                this.IsBusy = true;
+                this.LoadingMessage = "Building the type preview...";
+
+                var session = this.Sessions.Single(x => x.Assembler.Cache == this.SelectedIteration.Iteration.Cache);
+
+                var reqif = await Task.Run(() => new ReqIFBuilder().BuildReqIF(session, this.SelectedIteration.Iteration, this.IncludeDeprecated, selectedRequirementsSpecifications, this.SelectedFormat));
```
Comment (joao4all) [inline Requirements/ViewModels/ReqIF/ReqIfExportDialogViewModel.cs:420]: Why the task.run? Wouldn't be better to run it in the same thread?

File: Requirements/ViewModels/ReqIF/ReqIfImportDialogViewModel.cs [hunk 6]
```diff
@@ -302,7 +302,7 @@ private void ExecuteCancel()
         /// </summary>
         private void ExecuteBrowse()
         {
-            var result = this.fileDialogService.GetOpenFileDialog(true, true, false, "ReqIF files (*.reqif, *.xml)|*.reqif; *.xml|All Files (*.*)|*.*", ".reqif", this.Path, 1);
+            var result = this.fileDialogService.GetOpenFileDialog(true, true, false, "ReqIF files (*.reqif, *.reqifz, *.zip)|*.reqif; *.reqifz; *.zip|All Files (*.*)|*.*", ".reqif", this.Path, 1);
```
Comment (joao4all) [inline Requirements/ViewModels/ReqIF/ReqIfImportDialogViewModel.cs]: Just a question to make sure: previously the code was using .xml files too. Now it doesn't. Is this intended?

**Review by joao4all — CHANGES_REQUESTED** [review]

File: Requirements.Tests/ReqIF/ParameterTypeMappingDialogTestFixture.cs (hunk 1 above)
Comment (rdevoogtatstariongroup) [reply Requirements.Tests/ReqIF/ParameterTypeMappingDialogTestFixture.cs]: Fixed it

File: Requirements/ReqIFDal/ReqIFBuilder.cs (hunk 2 above)
Comment (rdevoogtatstariongroup) [reply Requirements/ReqIFDal/ReqIFBuilder.cs]: moved it to its own file

File: Requirements/ViewModels/ReqIF/ReqIfExportDialogViewModel.cs (hunk 4 above)
Comment (rdevoogtatstariongroup) [reply Requirements/ViewModels/ReqIF/ReqIfExportDialogViewModel.cs]: added them

File: Requirements/ViewModels/ReqIF/ReqIfExportDialogViewModel.cs (hunk 5 above)
Comment (rdevoogtatstariongroup) [reply Requirements/ViewModels/ReqIF/ReqIfExportDialogViewModel.cs:420]: This keeps the UI responsive and able to cancel

File: Requirements/ViewModels/ReqIF/ReqIfImportDialogViewModel.cs (hunk 6 above)
Comment (rdevoogtatstariongroup) [reply Requirements/ViewModels/ReqIF/ReqIfImportDialogViewModel.cs]: Restored *.xml

File: Requirements/ReqIFDal/ThingFactory.cs (hunk 3 above)
Comment (rdevoogtatstariongroup) [reply Requirements/ReqIFDal/ThingFactory.cs]: Made ConvertXhtmlToText private and rewrote the test to exercise it through the public ComputeRequirementThings path (an XHTML requirement value asserted on the resulting Requirement.Definition); on the other two points I kept regex rather than XmlReader because DOORS/Capella markup isn't guaranteed to be well-formed XML (an XmlReader would throw), but hoisted the patterns into static readonly RegexOptions.Compiled fields — [GeneratedRegex] source generators aren't available on net48 — and it now drops <script>/<style> with their content, covered by the edge-case assertions.

**Review by joao4all — APPROVED** [review]
